db/versions/0017.yml

```diff
@@ -0,0 +1,699 @@
+      If the pagination arguments are both NULL, all rows are returned.
+      Otherwise, page_size rows are returned at offset page_offset.
+    mode: read
+    serviceName: index
+    args: namespace_in text, name_in text, page_size_in integer, page_offset_in integer
+    returns: table(namespace text, name text, rank integer, task_id text, data jsonb, expires timestamptz)
+    body: |-
+      begin
+        return query
+        select
+          indexed_tasks.namespace,
+          indexed_tasks.name,
+          indexed_tasks.rank,
+          indexed_tasks.task_id,
+          indexed_tasks.data,
+          indexed_tasks.expires
+        from indexed_tasks
+        where
+          (indexed_tasks.namespace = namespace_in or namespace_in is null) and
+          (indexed_tasks.name = name_in or name_in is null) and
+          (indexed_tasks.expires > now())
+        -- we previously used to order by the hashed namespace but there's probably no need to add this complication.
+        order by indexed_tasks.namespace, name
+        limit get_page_limit(page_size_in)
+        offset get_page_offset(page_offset_in);
+      end
+  create_indexed_task:
+    description: |-
+      Create a new indexed task. Raises UNIQUE_VIOLATION if the indexed task already exists.
+      Returns the etag of the newly created indexed task.
+    mode: write
+    serviceName: index
+    args: namespace_in text, name_in text, rank_in integer, task_id_in text, data_in jsonb, expires_in timestamptz
+    returns: uuid
+    body: |-
+      declare
+        new_etag uuid := public.gen_random_uuid();
+      begin
+        insert
+          into indexed_tasks (namespace, name, rank, task_id, data, expires, etag)
+          values (namespace_in, name_in, rank_in, task_id_in, data_in, expires_in, new_etag);
+        return new_etag;
+      end
+  get_indexed_task:
+    description: |-
+      Get a non-expired indexed task. The returned table will have one or
+      (if no such indexed task is defined) an exception will be raised with error code 'P0002'.
+    mode: read
+    serviceName: index
+    args: namespace_in text, name_in text
+    returns: table(namespace text, name text, rank integer, task_id text, data jsonb, expires timestamptz, etag uuid)
+    body: |-
+      declare
+        new_row indexed_tasks%ROWTYPE;
+      begin
+        select
+          indexed_tasks.namespace,
+          indexed_tasks.name,
+          indexed_tasks.rank,
+          indexed_tasks.task_id,
+          indexed_tasks.data,
+          indexed_tasks.expires,
+          indexed_tasks.etag
+        from indexed_tasks
+        where
+          indexed_tasks.namespace = namespace_in and
+          indexed_tasks.name = name_in and
+          indexed_tasks.expires > now()
+        into new_row;
+        if found then
+          return query select new_row.namespace, new_row.name, new_row.rank, new_row.task_id, new_row.data, new_row.expires, new_row.etag;
+          return;
```

Comment:
> Why the double select here?

I'm not sure how to write it in a single select and make sure a "no such row error" is thrown when the row is not there.

> And since this returns a table (and thus an array of rows in JS) anyway, why not just return an empty set on not found?

The reason for that were:
* less refactoring of the index service to reduce the chance of introducing a bug
* we throw a 'no such row' error in `update_{worker|indexed_task}` when the row doesn't exist rather than return an empty row so I figured there shouldn't be an issue to extend this to this db function.

Thoughts?

Reply:
For the other `get_` methods, we've just returned the select and included a check on the length of the returned value:
https://github.com/taskcluster/taskcluster/blob/6c871ced2f2ebec3751c368b7e2cc459a799f3c2/services/worker-manager/src/data.js#L45-L47
That ends up with a simpler `get_` function, and the JS is a little shorter too (no `try/catch/err.code===`).